File: 2、LBOCExpandFriendList/LBExpandFriendList/折叠的表示图/LBExpandTableView.m
//
//  LBExpandTableView.m
//  LBExpandFriendList
//
//  Created by yunmei on 2018/1/19.
//  Copyright © 2018年 yunmei. All rights reserved.
//

#import "LBExpandTableView.h"
#import "LBExpandCell.h"
#import "UIColor+colorWithHex.h"
#define LBCellHeight 44
@implementation LBExpandTableView

-(instancetype)initWithFrame:(CGRect)frame style:(UITableViewStyle)style{
    
    self = [super initWithFrame:frame style:style];
    if (self) {
        self.delegate = self;
        self.dataSource = self;
    }
    return self;
}
-(void)setSectionTitleArray:(NSMutableArray *)sectionTitleArray{
    _sectionTitleArray = sectionTitleArray;
    [self reloadData];
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return _dataSourceArray.count;
}
- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    return _sectionTitleArray[section];
}
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    LBExpandCell *cell = [LBExpandCell cellWithTableView:tableView];
    cell.titleLabel.text = [NSString stringWithFormat:@"%ld组 %@",(long)indexPath.section,_dataSourceArray[indexPath.section][indexPath.row]];
    return cell;
}

- (NSInteger)tableView:(nonnull UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    if ([_sectionStateArray[section] isEqualToString:@"1"]){
        /// 组展开状态
        NSArray *array = [_dataSourceArray objectAtIndex:section];
        return array.count;
    }else{
        /// 组闭合状态,返回0
        return 0;
    }
}

- (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section
{   /// 添加按钮到组头视图
    UIButton *button = [UIButton buttonWithType:UIButtonTypeCustom];
    [button setFrame:CGRectMake(0, 0, self.frame.size.width, LBCellHeight)];
    button.tag = section + 1;
    button.backgroundColor = [UIColor whiteColor];
    [button setTitleColor:[UIColor grayColor] forState:UIControlStateNormal];
    [button setTitleEdgeInsets:UIEdgeInsetsMake(0, 0, 0, 60)];
    [button addTarget:self action:@selector(sectionButtonClick:) forControlEvents:UIControlEventTouchUpInside];
    /// 添加底部线
    UIImageView *line = [[UIImageView alloc]init];
    [line setImage:[UIImage imageNamed:@"line"]];
    [button addSubview:line];
    [line mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.equalTo(button);
        make.top.equalTo(button.mas_top).offset(LBCellHeight - 1);
        make.centerY.equalTo(button);
        make.width.equalTo(@([UIScreen mainScreen].bounds.size.width));
        make.height.equalTo(@1);
    }];
    /// 添加左边图标
    UIImageView *leftImgView = [[UIImageView alloc]init];
    [leftImgView setImage:[UIImage imageNamed:@"circle"]];
    [button addSubview:leftImgView];
    [leftImgView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.width.height.equalTo(@24);
        make.left.equalTo(button).offset(10);
        make.centerY.equalTo(button);
    }];
    /// 添加右边的箭头
    UIImageView *arrowImgView = [[UIImageView alloc]init];
    [button addSubview:arrowImgView];
    [arrowImgView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.width.height.equalTo(@10);
        make.centerY.equalTo(button);
        make.left.equalTo(button.mas_left).offset([UIScreen mainScreen].bounds.size.width - 30);
    }];
    if ([_sectionStateArray[section] isEqualToString:@"0"]) {
        arrowImgView.image = [UIImage imageNamed:@"bottomArrow"];
    }else if ([_sectionStateArray[section] isEqualToString:@"1"]) {
        arrowImgView.image = [UIImage imageNamed:@"topArrow"];
    }
    /// 每组个数
    UILabel *countLabel = [[UILabel alloc]init];
    [countLabel setBackgroundColor:[UIColor clearColor]];
    countLabel.font = [UIFont systemFontOfSize:14];
    countLabel.text = [NSString stringWithFormat:@"%lu",(unsigned long)[_dataSourceArray[section] count]];
    [button addSubview:countLabel];
    [countLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.equalTo(button);
        make.right.equalTo(arrowImgView.mas_left).offset(- 20);
    }];
    /// 组标题
    UILabel *sectionTitleLabel = [[UILabel alloc]init];
    [sectionTitleLabel setBackgroundColor:[UIColor clearColor]];
    sectionTitleLabel.font = [UIFont systemFontOfSize:14];
    sectionTitleLabel.text = _sectionTitleArray[section];
    [button addSubview:sectionTitleLabel];
    [sectionTitleLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.equalTo(button);
        make.left.equalTo(leftImgView.mas_right).offset(15);
    }];
    return button;
}
- (void)sectionButtonClick:(UIButton *)button
{
    /// 判断点击组被点击的状态值
    if ([_sectionStateArray[button.tag - 1] isEqualToString:@"1"]){
        /// 修改点击的状态值
        [_sectionStateArray replaceObjectAtIndex:button.tag - 1 withObject:@"0"];
    }else{
        [_sectionStateArray replaceObjectAtIndex:button.tag - 1 withObject:@"1"];
    }
    /// 刷新组
    [self reloadSections:[NSIndexSet indexSetWithIndex:button.tag - 1] withRowAnimation:UITableViewRowAnimationAutomatic];
}
- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath{
    return LBCellHeight;
}
- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section{
    return LBCellHeight;
}
-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
}
@end

File: LBSwiftExpandFriendList/LBSwiftExpandFriendList/可折叠的表示图/LBExpandTableView.swift
//
//  LBExpandTableView.swift
//  LBSwiftExpandFriendList
//
//  Created by yunmei on 2018/1/19.
//  Copyright © 2018年 liubo. All rights reserved.
//

import UIKit
import SnapKit
let LBCellHeight = 44

class LBExpandTableView: UITableView {
    lazy var sectionTitleArray:[String] = [String]()
    lazy var sectionStateArray:[String] = [String]()
    lazy var dataSourceArray:[[String]] = [[String]]()
    override init(frame: CGRect, style: UITableViewStyle) {
        super.init(frame: frame, style: style)
        self.delegate = self
        self.dataSource = self
        self.register(LBExpandCell.self, forCellReuseIdentifier: "LBExpandCell")
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
}
extension LBExpandTableView:UITableViewDelegate,UITableViewDataSource{
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let index = sectionStateArray[section]
        
        if index == "1" {
            /// 组展开状态
            let array = dataSourceArray[section]
            return array.count
        }
        /// 组闭合状态,返回0
        return 0
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "LBExpandCell")as!LBExpandCell
        cell.titleLabel.text = "\(indexPath.section)组 \(dataSourceArray[indexPath.section][indexPath.row])"
        return cell;
    }
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return dataSourceArray.count
    }
    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        return sectionTitleArray[section]
    }
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return CGFloat(LBCellHeight)
    }
    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return CGFloat(LBCellHeight)
    }
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
    }
    func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        /// 添加按钮到组头视图
        let button = UIButton.init(type: .custom)
        button.frame = CGRect(x: 0, y: 0, width: Int(UIScreen.main.bounds.size.width), height: LBCellHeight)
        button.tag = section + 1
        button.backgroundColor = UIColor.white
        button.setTitleColor(UIColor.gray, for: .normal)
        button.addTarget(self, action: #selector(LBExpandTableView.sectionButtonClick(button:)), for: .touchUpInside)
        button.titleEdgeInsets = UIEdgeInsets.init(top: 0, left: 0, bottom: 0, right: 60)
        /// 添加底部线
        let line = UIImageView()
        line.image = UIImage.init(named: "line")
        button.addSubview(line)
        line.snp.makeConstraints { (make) in
            make.left.equalTo(button);
            make.top.equalTo(button.snp.top).offset(LBCellHeight - 1);
            make.centerY.equalTo(button);
            make.width.equalTo(UIScreen.main.bounds.size.width);
            make.height.equalTo(1);
        }
        /// 添加左边图标
        let leftImgView = UIImageView()
        leftImgView.image = UIImage.init(named: "circle")
        button.addSubview(leftImgView)
        leftImgView.snp.makeConstraints { (make) in
            make.width.height.equalTo(24);
            make.left.equalTo(button).offset(10);
            make.centerY.equalTo(button);
        }
        /// 添加右边的箭头
        let arrowImgView = UIImageView()
        button.addSubview(arrowImgView)
        arrowImgView.snp.makeConstraints { (make) in
            make.width.height.equalTo(10);
            make.centerY.equalTo(button);
            make.left.equalTo(button.snp.left).offset(UIScreen.main.bounds.size.width - 30);
        }
        let state = sectionStateArray[section]
        
        if state == "0" {
            arrowImgView.image = UIImage.init(named: "bottomArrow");
        }else if state == "1" {
            arrowImgView.image = UIImage.init(named: "topArrow")
        }
        /// 每组个数
        let countLabel = UILabel()
        countLabel.backgroundColor = UIColor.clear
        countLabel.font = UIFont.systemFont(ofSize: 14)
        let dataArray = dataSourceArray[section].count
        countLabel.text = "\(dataArray)"
        button.addSubview(countLabel)
        countLabel.snp.makeConstraints { (make) in
            make.centerY.equalTo(button);
            make.right.equalTo(arrowImgView.snp.left).offset(-20);
        }
        /// 组标题
        let sectionTitleLabel = UILabel()
        sectionTitleLabel.backgroundColor = UIColor.clear
        sectionTitleLabel.font = UIFont.systemFont(ofSize: 14)
        sectionTitleLabel.text = sectionTitleArray[section];
        button.addSubview(sectionTitleLabel)
        sectionTitleLabel.snp.makeConstraints { (make) in
            make.centerY.equalTo(button);
            make.left.equalTo(leftImgView.snp.right).offset(15);
        }

        return button;
    }
}
extension LBExpandTableView{
    @objc func sectionButtonClick(button:UIButton){
        
        /// 判断点击组被点击的状态值
        let state = sectionStateArray[button.tag - 1]
        
        if (state == "1"){
            /// 修改点击的状态值
            sectionStateArray[button.tag - 1] = "0"
        }else{
            sectionStateArray[button.tag - 1] = "1"
        }
        /// 刷新组
        let indexSet = NSIndexSet.init(index: button.tag - 1) as IndexSet
        reloadSections(indexSet, with: .automatic)
    }
}
